### PR TITLE
Add nodePorts option to multicluster helm chart

### PR DESCRIPTION
**Pull request by psmit**
For LoadBalancer and NodePort type services it should be possible to set
a specific node port, this wasn't possible yet.

This commit adds nodePorts to the multicluster helm chart

Signed-off-by: Peter Smit <peter.smit@inscripta.io>

### CODE REVIEW

**Comment by kleimkuhler** [comment]
Thanks for opening this up @psmit! You'll need to run `bin/helm-docs` to update the chart READMEs so that CI can pass.

I'm wondering if it makes sense to make a `gateway` record that has sub-records for each of the `gateway...` prefixed values right now—so `gatewayName` -> `gateway.name`. This would allow adding `gateway.nodePort.port` and `gateway.nodePort.probePort`. That way, it would be be more obvious that these values are specific to the gateway.

This would lead to a bigger change, but it does add some clarity that is currently not there; it may not be obvious that `nodePorts` value is specific to the gateway.

**Comment by psmit** [comment]
Thanks for the feedback @kleimkuhler . I have now grouped the variables under `gateway:`, but this would be a breaking change. Is that ok?

**Comment by psmit** [comment]
As far as I can see, none of the integration test failures have anything to do with the multicluster chart. Can someone confirm this?

**Comment by alpeb** [comment]
Given the structure of the `values.yaml` file was changed, its go counterpart `multicluster/values.values.go` and the other places in go code were this struct is referred to need to be updated. Glad to provide some additional guidance on that if needed :wink:

**Comment by psmit** [comment]
@alpeb Thanks for the hint. This is indeed new for me, but I think I'll be able to adjust this accordingly. Could you give me a hint on testing this locally before I push a new commit?

**Comment by alpeb** [comment]
@psmit in order to test this you have to install multicluster in your target cluster with something like this:
```
linkerd mc install --set gateway.serviceType=NodePort --set gateway.nodePort=30500 --set gateway.probe.nodePort=30501|k apply -f -
```
And then generate the Link resources with
```
linkerd mc link --cluster-name target --set gateway.probe.port=30501
```
I realized this still requires additional changes. `multicluster/charts/linkerd-multicluster-link/values.yaml` also uses the `Gateway` struct in `multicluster/values/values.go`. So you need to modify `gateway` in that file to also have a tree structure.

On the other hand, `extractGatewayPort()` in `multicluster/cmd/gateways.go` (used when calling `linkerd mc link`) extracts the target cluster gateway port checking at the `Port` field in the gateway service's `spec.Ports` entries. You have to modify this to also check for `NodePort` if you're using that service type.

Separate question: The CI runs on this PR are requesting manual approval, which only happens when files under `.github/workflows` are modified. Did you happen to have done that, perhaps by mistake, in one of your earlier commits?

**Comment by psmit** [comment]
> I realized this still requires additional changes. `multicluster/charts/linkerd-multicluster-link/values.yaml` also uses the `Gateway` struct in `multicluster/values/values.go`. So you need to modify `gateway` in that file to also have a tree structure.
 
This I already had done. One question there is; do we keep the cli-flags the same or do we look to have dot-syntax in there as well?

> On the other hand, `extractGatewayPort()` in `multicluster/cmd/gateways.go` (used when calling `linkerd mc link`) extracts the target cluster gateway port checking at the `Port` field in the gateway service's `spec.Ports` entries. You have to modify this to also check for `NodePort` if you're using that service type.

In my own case, the NodePort would not correspond to the actual public port / IP, because there is a load balancer (not managed in k8s) before the service. I think this would be the primary use case of using the NodePort type service.

> Separate question: The CI runs on this PR are requesting manual approval, which only happens when files under `.github/workflows` are modified. Did you happen to have done that, perhaps by mistake, in one of your earlier commits?

No, I have not touched that folder. For me it says: `First-time contributors need a maintainer to approve running workflows`

**Review by alpeb — COMMENTED** [review]
> > I realized this still requires additional changes. `multicluster/charts/linkerd-multicluster-link/values.yaml` also uses the `Gateway` struct in `multicluster/values/values.go`. So you need to modify `gateway` in that file to also have a tree structure.
> 
> This I already had done. One question there is; do we keep the cli-flags the same or do we look to have dot-syntax in there as well?

You modified `multicluster/charts/linkerd-multicluster/values.yaml` but I don't see changes to `multicluster/charts/linkerd-multicluster-link/values.yaml`. Note this also requires updating `gateway-mirror.yaml` to make reference to `{{.Values.gateway.probe.port}}`. As for the CLI flags, let's leave them as-is. This is not such a common scenario, so it'd be fine to access the additional config through helm-like flags (`--set` and friends).

> > On the other hand, `extractGatewayPort()` in `multicluster/cmd/gateways.go` (used when calling `linkerd mc link`) extracts the target cluster gateway port checking at the `Port` field in the gateway service's `spec.Ports` entries. You have to modify this to also check for `NodePort` if you're using that service type.
> 
> In my own case, the NodePort would not correspond to the actual public port / IP, because there is a load balancer (not managed in k8s) before the service. I think this would be the primary use case of using the NodePort type service.

Ok that'd mean the LB would have to listen on port `gateway.port` and forward that to `gateway.nodePort`. But I think that when setting the service type to NodePort people should expect it to work without any additional setup. We could introduce extra config to have the Link's target `gatewayPort` be different than the (Node)port the gateway service is actually listening to, but IMO it's not worth complicating things. The easiest way should be to expand the mechanism I pointed above to take into account NodePort services. And so if an LB is used to front the NodePort, it'd listen to the same port it'll be forwarding to... Let me know if this makes sense or I'm missing something.

**Comment by psmit** [comment]
Thanks @alpeb for your extensive review. I have tried to address all your comments, but I'm a small bit "in over my head". Originally I thought that this was simple as only changing the helm chart; not realizing that there was a lot of matching Go code that needed updating as well.

File: multicluster/values/values.go [hunk 1]
```diff
@@ -18,29 +18,43 @@ const (
...
+	ServiceMirrorRetryLimit        uint32   `json:"serviceMirrorRetryLimit"`
+	ServiceMirrorUID               int64    `json:"serviceMirrorUID"`
+	RemoteMirrorServiceAccount     bool     `json:"remoteMirrorServiceAccount"`
+	RemoteMirrorServiceAccountName string   `json:"remoteMirrorServiceAccountName"`
+	TargetClusterName              string   `json:"targetClusterName"`
+}
+
+// Gateway contains all opttions related to the Gateway Service
+type Gateway struct {
+	Enabled            bool              `json:"bool"`
```
Comment (alpeb) [inline multicluster/values/values.go:42]:
```suggestion
	Enabled            bool              `json:"enabled"`
```

**Review by alpeb — COMMENTED** [review]
Thanks again @psmit , you're doing great, we're almost here :-)
The integration tests are failing because of the error I pointed below. I'll be giving this another deep try tomorrow and let you know :wink:

**Review by alpeb — COMMENTED** [review]
Thanks again @psmit , you're doing great, we're almost there :-)
The integration tests are failing because of the error I pointed below. I'll be giving this another deep try tomorrow and let you know :wink:

File: multicluster/charts/linkerd-multicluster/templates/gateway.yaml [hunk 2]
```diff
@@ -10,64 +10,71 @@ metadata:
...
 metadata:
-  name: {{.Values.gatewayName}}
+  name: {{.Values.gateway.name}}
   namespace: {{.Values.namespace}}
   labels:
     linkerd.io/extension: multicluster
   annotations:
-    mirror.linkerd.io/gateway-identity: {{.Values.gatewayName}}.{{.Values.namespace}}.serviceaccount.identity.{{.Values.linkerdNamespace}}.{{.Values.identityTrustDomain}}
+    mirror.linkerd.io/gateway-identity: {{.Values.gateway.name}}.{{.Values.namespace}}.serviceaccount.identity.{{.Values.linkerdNamespace}}.{{.Values.identityTrustDomain}}
     mirror.linkerd.io/probe-period: "{{.Values.gatewayProbeSeconds}}"
```
Comment (alpeb) [inline multicluster/charts/linkerd-multicluster/templates/gateway.yaml:46]:
```suggestion
    mirror.linkerd.io/probe-period: "{{.Values.gateway.probe.seconds}}"
```

**Comment by psmit** [comment]
Thanks for the help! I'm at this moment stuck locally with a nil pointer reference, reading `values.Proxy.Ports.Outbound` in `multicluster/cmd/install.go`:

```
bin/go-run cli mc --context=se02 install --set gateway.serviceType=NodePort --set gateway.nodePort=30500 --set gateway.probe.nodePort=30501
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x23fef15]

goroutine 1 [running]:
github.com/linkerd/linkerd2/multicluster/cmd.buildMulticlusterInstallValues(0x2ba0080, 0xc00013e010, 0xc00040bc70, 0x0, 0x0, 0x0)
	/work/linkerd2/multicluster/cmd/install.go:220 +0x695
github.com/linkerd/linkerd2/multicluster/cmd.newMulticlusterInstallCommand.func1(0xc000488dc0, 0xc00040bf80, 0x0, 0x7, 0x0, 0x0)
	/work/linkerd2/multicluster/cmd/install.go:72 +0x310
github.com/spf13/cobra.(*Command).execute(0xc000488dc0, 0xc00040bf10, 0x7, 0x7, 0xc000488dc0, 0xc00040bf10)
	/home/psmit/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:842 +0x506
github.com/spf13/cobra.(*Command).ExecuteC(0x3845be0, 0xc000144010, 0x9, 0x9)
	/home/psmit/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:950 +0x4b3
github.com/spf13/cobra.(*Command).Execute(...)
	/home/psmit/go/pkg/mod/github.com/spf13/cobra@v1.0.0/command.go:887
main.main()
	/work/linkerd2/cli/main.go:34 +0x236
```
	
I have tried to debug through the code and it seems that this happens because in healthcheck.go:1722 this value is removed from the yaml file using `RemoveGlobalFieldIfPresent`. I have now idea why this happens or how to fix it (or if it is even related to the code changes I made now).

**Comment by alpeb** [comment]
That last problem sounds unrelated to this... maybe you have an old `linkerd-config` ConfigMap in your cluster? Can you post its (redacted) contents here?

On a separate note, I think you missed [this suggestion](https://github.com/linkerd/linkerd2/pull/6059#discussion_r626164552) .

Also, `extractPort()` in `pkg/multicluster/link.go` requires a similar change to what you did with `extractGatewayPort()` in order to extract the NodePort for the probe. The `port` argument in that function will have to be replaced with `gateway` so you can check  for `gateway.Spec.Type == "NodePort"` like you did in the other function.

Finally, since NodePort services don't expose the node's IP in their status, it has to be provided explicitly through the `--gateway-addresses` option when doing `linkerd mc link`. This is fine, but there's not a way to set that when installing through Helm, so I raised #6090, so that can be tackled later (I could take a stab at that one when this one gets merged).

**Comment by psmit** [comment]
Hi @alpeb. I have tried to install multicluster on a fresh install of linkerd edge 21.4.5. My `linkerd-configmap` looks like this:

```
apiVersion: v1
data:
  values: |
    clusterDomain: cluster.local
    clusterNetworks: 10.0.0.0/8,100.64.0.0/10,172.16.0.0/12,192.168.0.0/16
    cniEnabled: false
    controlPlaneTracing: false
    controlPlaneTracingNamespace: linkerd-jaeger
    controllerImage: cr.l5d.io/linkerd/controller
    controllerLogFormat: plain
    controllerLogLevel: info
    controllerReplicas: 2
    controllerResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    controllerUID: 2103
    debugContainer:
      image:
        name: cr.l5d.io/linkerd/debug
        pullPolicy: ""
        version: ""
    destinationResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    disableHeartBeat: false
    enableEndpointSlices: false
    enableH2Upgrade: true
    enablePodAntiAffinity: true
    global:
      proxy:
        resources:
          cpu:
            limit: "1"
            request: 10m
          memory:
            limit: 250Mi
            request: 20Mi
    grafanaResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 1024Mi
        request: 30Mi
    heartbeatResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    identity:
      issuer:
        clockSkewAllowance: 20s
        crtExpiry: null
        issuanceLifetime: 24h0m0s
        scheme: kubernetes.io/tls
        tls:
          crtPEM: ""
    identityResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 10Mi
    identityTrustAnchorsPEM: |
      -----BEGIN CERTIFICATE-----
      REDACTED
      -----END CERTIFICATE-----
    identityTrustDomain: cluster.local
    imagePullPolicy: IfNotPresent
    imagePullSecrets: []
    installNamespace: false
    linkerdVersion: edge-21.4.5
    namespace: linkerd
    nodeSelector:
      beta.kubernetes.io/os: linux
    omitWebhookSideEffects: false
    podAnnotations:
      cluster-autoscaler.kubernetes.io/safe-to-evict: "true"
    podLabels: {}
    profileValidator:
      caBundle: ""
      crtPEM: ""
      externalSecret: false
      namespaceSelector:
        matchExpressions:
        - key: config.linkerd.io/admission-webhooks
          operator: NotIn
          values:
          - disabled
    prometheusResources:
      cpu:
        limit: "4"
        request: 20m
      memory:
        limit: 8192Mi
        request: 300Mi
    proxy:
      await: true
      cores: 0
      enableExternalProfiles: false
      image:
        name: cr.l5d.io/linkerd/proxy
        pullPolicy: ""
        version: ""
      inboundConnectTimeout: 100ms
      logFormat: plain
      logLevel: warn,linkerd=info
      opaquePorts: 25,443,587,3306,5432,11211
      outboundConnectTimeout: 1000ms
      ports:
        admin: 4191
        control: 4190
        inbound: 4143
        outbound: 4140
      requireIdentityOnInboundPorts: ""
      resources:
        cpu:
          limit: ""
          request: ""
        memory:
          limit: ""
          request: ""
      uid: 2102
      waitBeforeExitSeconds: 0
    proxyInit:
      closeWaitTimeoutSecs: 0
      ignoreInboundPorts: ""
      ignoreOutboundPorts: ""
      image:
        name: cr.l5d.io/linkerd/proxy-init
        pullPolicy: ""
        version: v1.3.11
      resources:
        cpu:
          limit: 100m
          request: 10m
        memory:
          limit: 50Mi
          request: 10Mi
      xtMountPath:
        mountPath: /run
        name: linkerd-proxy-init-xtables-lock
    proxyInjector:
      caBundle: ""
      crtPEM: ""
      externalSecret: false
      namespaceSelector:
        matchExpressions:
        - key: config.linkerd.io/admission-webhooks
          operator: NotIn
          values:
          - disabled
    proxyInjectorResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    publicAPIResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    spValidatorResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    webResources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 30Mi
    webhookFailurePolicy: Fail
kind: ConfigMap
metadata:
  annotations:
    linkerd.io/created-by: linkerd/helm edge-21.4.5
    meta.helm.sh/release-name: linkerd2
    meta.helm.sh/release-namespace: linkerd
  creationTimestamp: "2021-05-07T04:37:31Z"
  labels:
    app.kubernetes.io/managed-by: Helm
    linkerd.io/control-plane-component: controller
    linkerd.io/control-plane-ns: linkerd
  name: linkerd-config
  namespace: linkerd
  resourceVersion: "11406347"
  selfLink: /api/v1/namespaces/linkerd/configmaps/linkerd-config
  uid: 47e2521a-e1cf-44a6-9450-400b22765bb8
```

**Comment by alpeb** [comment]
@psmit your last changes are working great, and as you can see, CI is green :+1: 
As for the error you're seeing, I see your config does have a value for `proxy.ports.outbound` so I don't understand why it's failing, nor can I reproduce it... it could be that `getLinkerdConfigMap()` isn't able to retrieve `values` at all. Some extra debugging in that function might give you some clues.

**Comment by psmit** [comment]
All right, I have been trying to debug this; and I noticed the following:

- I'm already getting this errors with previous linkerd versions. I also can't install with `linkerd mc install` on 2.10.1
- I'm seeing this behavior consistently and on multiple clusters
- These clusters all (basically) use the same values.yaml

Long debugging later, it turns out that:
```
global:
  proxy:
    resources:
      cpu:
        limit: "1"
        request: 10m
      memory:
        limit: 250Mi
        request: 20Mi
```

Is the culprit. If this snippet is in my values.yaml, then `RemoveGlobalFieldIfPresent` also removes the `proxy` map, including the ports...

That said; I think I should raise a separate issue for this? There should now not be anything blocking for reviewing this MR further?

**Review by kleimkuhler — APPROVED** [review]
This looks great @psmit thank you so much for getting these additional changes in.

**Pull request merged**